### PR TITLE
ToGlib should take Self

### LINKED ISSUE #410: `ToGlib` trait should take `self` instead of `&self`
Somehow I'm getting those with clippy now

```
error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:   --> sourceview5/src/auto/enums.rs:45:16
   |
45 |     fn to_glib(&self) -> ffi::GtkSourceBackgroundPatternType {
   |                ^^^^^
   |
   = note: `-D clippy::wrong-self-convention` implied by `-D warnings`
   = help: consider choosing a less ambiguous name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/enums.rs:126:16
    |
126 |     fn to_glib(&self) -> ffi::GtkSourceBracketMatchType {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/enums.rs:211:16
    |
211 |     fn to_glib(&self) -> ffi::GtkSourceChangeCaseType {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/enums.rs:293:16
    |
293 |     fn to_glib(&self) -> ffi::GtkSourceCompletionActivation {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/enums.rs:384:16
    |
384 |     fn to_glib(&self) -> ffi::GtkSourceCompletionColumn {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

Error:   --> sourceview5/src/auto/flags.rs:33:16
   |
33 |     fn to_glib(&self) -> ffi::GtkSourceFileSaverFlags {
   |                ^^^^^
   |
   = help: consider choosing a less ambiguous name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/flags.rs:204:16
    |
204 |     fn to_glib(&self) -> ffi::GtkSourceSpaceTypeFlags {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:   --> sourceview5/src/auto/flags.rs:89:16
   |
89 |     fn to_glib(&self) -> ffi::GtkSourceSortFlags {
   |                ^^^^^
   |
   = help: consider choosing a less ambiguous name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/flags.rs:146:16
    |
146 |     fn to_glib(&self) -> ffi::GtkSourceSpaceLocationFlags {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

error: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
Error:    --> sourceview5/src/auto/flags.rs:204:16
    |
204 |     fn to_glib(&self) -> ffi::GtkSourceSpaceTypeFlags {
    |                ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
```

related to https://github.com/gtk-rs/gtk-rs/issues/313

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Fixes #410 

Requires https://github.com/gtk-rs/gir/pull/1093
- [x] gtk4-rs https://github.com/gtk-rs/gtk4-rs/pull/331
- [x] gstreamer-rs https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/merge_requests/758
- [x] gst-plugins-rs https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/merge_requests/507

### CODE REVIEW

**Comment by sdroege** [comment]
This breaks the `ToGlib` impl for `SignalHandlerId` and similar cases

**Comment by bilelmoussaoui** [comment]
> This breaks the `ToGlib` impl for `SignalHandlerId` and similar cases

/me blames Sublime Text's search being so slow that I thought there were no search results left. Should be handled now

**Comment by sdroege** [comment]
I mean `SignalHandlerId` and similar are not `Copy` :)

**Comment by sdroege** [comment]
> glib: mark SourceId/SignalHandlerId copy

I mean they are intentionally *not* `Clone` or `Copy` :)

**Comment by sdroege** [comment]
> I mean they are intentionally _not_ `Clone` or `Copy` :)

How should this be handled? Have them not implement `ToGlib`?

**Comment by bilelmoussaoui** [comment]
I have no idea what's causing 
```
error[E0277]: the trait bound `&Animal: ToGlib` is not satisfied
  --> glib-macros/tests/test.rs:27:49
   |
27 |     #[derive(Debug, Eq, PartialEq, Clone, Copy, GEnum)]
   |                                                 ^^^^^ the trait `ToGlib` is not implemented for `&Animal`
   |
   = help: the following implementations were found:
             <Animal as ToGlib>
   = note: required by `to_glib`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `&MyFlags: ToGlib` is not satisfied
   --> glib-macros/tests/test.rs:115:5
    |
115 |     #[gflags("MyFlags")]
    |     ^^^^^^^^^^^^^^^^^^^^ the trait `ToGlib` is not implemented for `&MyFlags`
    |
    = help: the following implementations were found:
              <MyFlags as ToGlib>
    = note: required by `to_glib`
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by bilelmoussaoui** [comment]
> > I mean they are intentionally _not_ `Clone` or `Copy` :)
> 
> How should this be handled? Have them not implement `ToGlib`?

Ah indeed. I guess we can just not implement ToGlib for them, there are only two of them :)

**Comment by sdroege** [comment]
> Ah indeed. I guess we can just not implement ToGlib for them, there are only two of them :)

Are there any other types where you had to add `Clone` / `Copy`?

> I have no idea what's causing
> [...]

The enum/flags macros should implement the traits on `T` instead of `&T`, use them as such and the types themselves need to derive `Clone` and `Copy`.

**Comment by bilelmoussaoui** [comment]
> Are there any other types where you had to add `Clone` / `Copy`?

There were the only types, I ended up adding a `pub unsafe fn as_raw(&self) -> T`

> 
> > I have no idea what's causing
> > [...]
> 
> The enum/flags macros should implement the traits on `T` instead of `&T`, use them as such and the types themselves need to derive `Clone` and `Copy`.

The issue comes from the `SetValue` trait I think where we get a `&Self` that doesn't implement `to_glib()`. Trying to clone &Self to get Self back gives me a 
```
error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
  --> glib-macros/tests/test.rs:27:49
   |
27 |     #[derive(Debug, Eq, PartialEq, Clone, Copy, GEnum)]
   |                                                 ^^^^^
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
   --> glib-macros/tests/test.rs:115:5
    |
115 |     #[gflags("MyFlags")]
    |     ^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```
which really doesn't help ^^

**Comment by sdroege** [comment]
Can you give the output with `-Z macro-backtrace`? That probably makes more sense

**Comment by bilelmoussaoui** [comment]
Here's the output with `-Z macro-backtrace`
```
error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
   --> glib-macros/tests/test.rs:27:49
    |
27  |     #[derive(Debug, Eq, PartialEq, Clone, Copy, GEnum)]
    |                                                 ^^^^^ in this macro invocation
    | 
   ::: /home/bilelmoussaoui/Projects/gtk-rs/glib-macros/src/lib.rs:247:1
    |
247 | pub fn genum_derive(input: TokenStream) -> TokenStream {
    | ------------------------------------------------------ in this expansion of `#[derive(GEnum)]`

error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
   --> glib-macros/tests/test.rs:115:5
    |
115 |     #[gflags("MyFlags")]
    |     ^^^^^^^^^^^^^^^^^^^^ in this macro invocation
    | 
   ::: /home/bilelmoussaoui/Projects/gtk-rs/glib-macros/src/lib.rs:335:1
    |
335 | pub fn gflags(attr: TokenStream, item: TokenStream) -> TokenStream {
    | ------------------------------------------------------------------ in this expansion of `#[gflags]`

error: aborting due to 2 previous errors
```
It doesn't give any useful information sadly :/

**Comment by GuillaumeGomez** [comment]
Time to give a try to `cargo expand` then.

**Comment by bilelmoussaoui** [comment]
> Time to give a try to `cargo expand` then.

I ended up doing that & fixed the issue, next time I should stop being lazy.

**Comment by sdroege** [comment]
The expanded code from the macro must be calling `to_glib()` or similar in a constant scope somewhere.

**Comment by bilelmoussaoui** [comment]
> The expanded code from the macro must be calling to_glib() or similar in a constant scope somewhere.

It was exactly that :-)

File: glib-macros/src/genum_derive.rs [hunk 1]
```diff
@@ -133,7 +133,7 @@ pub fn impl_genum(input: &syn::DeriveInput) -> TokenStream {
             unsafe fn set_value(value: &mut #crate_ident::Value, this: &Self) {
                 #crate_ident::gobject_ffi::g_value_set_enum(
                     #crate_ident::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
-                    #crate_ident::translate::ToGlib::to_glib(this))
+                    #crate_ident::translate::ToGlib::to_glib(this.clone()))
```
Comment (sdroege) [inline glib-macros/src/genum_derive.rs:136]: The `clone()` shouldn't be needed here: the type is `Copy` after all

File: glib-macros/src/gflags_attribute.rs [hunk 2]
```diff
@@ -152,8 +152,7 @@ pub fn impl_gflags(input: &DeriveInput, gtype_name: &LitStr) -> TokenStream {
         impl #crate_ident::value::SetValue for #name {
             unsafe fn set_value(value: &mut #crate_ident::Value, this: &Self) {
                 #crate_ident::gobject_ffi::g_value_set_flags(
-                    #crate_ident::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
-                    #crate_ident::translate::ToGlib::to_glib(this))
```
Comment (sdroege) [inline glib-macros/src/gflags_attribute.rs:156]: This shouldn't be necessary. The type should be `Copy` so you can use `to_glib()` here as before.

File: glib-macros/stuff.tst [hunk 3]
```diff
@@ -0,0 +1,6681 @@
+#![feature(prelude_import)]
```
Comment (sdroege) [inline glib-macros/stuff.tst:1]: This file probably shouldn't be here

**Review by sdroege — COMMENTED** [review]
Otherwise good to be merged

File: glib-macros/stuff.tst (hunk 3 above)
Comment (bilelmoussaoui) [reply glib-macros/stuff.tst:1]: Oups.

**Comment by bilelmoussaoui** [comment]
Clippy doesn't seem to like `to_glib` as a method name, maybe we can rename it to `into_glib`?

**Comment by sdroege** [comment]
> Clippy doesn't seem to like `to_glib` as a method name, maybe we can rename it to `into_glib`?

That would be the correct naming in this case, yes.

**Comment by bilelmoussaoui** [comment]
Should the trait itself be renamed to `IntoGlib` as well?

**Comment by sdroege** [comment]
That would make sense

**Comment by pbor** [comment]
We now have `into_raw` for `Value` and for `Error`. Should these be changed to `into_glib`?

**Comment by bilelmoussaoui** [comment]
This should be rebased and ready. I will open a corresponding PR against gtk4-rs/gstreamer-rs as well a bit later today, don't merge before that :)

**Comment by bilelmoussaoui** [comment]
> We now have `into_raw` for `Value` and for `Error`. Should these be changed to `into_glib`?

I don't think so, they are not Copy types.

**Comment by bilelmoussaoui** [comment]
@sdroege this one is ready to go :)

**Pull request merged**